File: vapor-app/src/main/java/com/tevinjeffrey/vapor/ui/files/fragments/MVPFragment.java
package com.tevinjeffrey.vapor.ui.files.fragments;

import android.os.Bundle;
import android.support.v4.app.Fragment;

import com.tevinjeffrey.vapor.ui.base.Presenter;
import com.tevinjeffrey.vapor.ui.base.View;
import com.tevinjeffrey.vapor.ui.files.FilesActivity;

import butterknife.ButterKnife;
import icepick.Icepick;

public abstract class MVPFragment extends Fragment implements View {

    Presenter<View> mPresenter;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Icepick.restoreInstanceState(this, savedInstanceState);
    }

    @Override
    public void onResume() {
        super.onResume();
        if (mPresenter != null) {
            mPresenter.onResume();
        }
    }

    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        if (mPresenter != null) {
            mPresenter.onActivityCreated(savedInstanceState);
        }
    }

    @Override
    public void onPause() {
        super.onPause();
        if (mPresenter != null) {
            mPresenter.onPause();
        }
    }


    @Override
    public void onDestroyView() {
        super.onDestroyView();
        ButterKnife.unbind(this);
        if (mPresenter != null) {
            mPresenter.detachView();
        }
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        Icepick.saveInstanceState(this, outState);
    }


    FilesActivity getParentActivity() {
        return (FilesActivity) getActivity();
    }

    @Override
    public String toString() {
        return this.getClass().getSimpleName();
    }

}

File: vapor-app/src/main/java/com/tevinjeffrey/vapor/ui/files/fragments/MVPFragment.kt
package com.tevinjeffrey.vapor.ui.files.fragments

import android.os.Bundle
import android.support.v4.app.Fragment

import com.tevinjeffrey.vapor.ui.base.Presenter
import com.tevinjeffrey.vapor.ui.base.View
import com.tevinjeffrey.vapor.ui.files.FilesActivity

import butterknife.ButterKnife
import com.tevinjeffrey.vapor.ui.base.BasePresenter
import com.tevinjeffrey.vapor.ui.files.FilesActivityPresenterImpl
import com.tevinjeffrey.vapor.ui.files.fragments.presenters.FilesPresenter
import icepick.Icepick

abstract class MVPFragment : Fragment(), View {

    var mPresenter: FilesPresenter<FilesFragmentView>? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        Icepick.restoreInstanceState(this, savedInstanceState)
    }

    override fun onResume() {
        super.onResume()
        if (mPresenter != null) {
            mPresenter!!.onResume()
        }
    }

    override fun onActivityCreated(savedInstanceState: Bundle) {
        super.onActivityCreated(savedInstanceState)
        if (mPresenter != null) {
            mPresenter!!.onActivityCreated(savedInstanceState)
        }
    }

    override fun onPause() {
        super.onPause()
        if (mPresenter != null) {
            mPresenter!!.onPause()
        }
    }


    override fun onDestroyView() {
        super.onDestroyView()
        ButterKnife.unbind(this)
        if (mPresenter != null) {
            mPresenter!!.detachView()
        }
    }

    override fun onSaveInstanceState(outState: Bundle?) {
        super.onSaveInstanceState(outState)
        Icepick.saveInstanceState(this, outState)
    }


    val parentActivity: FilesActivity
        get() = activity as FilesActivity

    override fun toString(): String {
        return this.javaClass.simpleName
    }

}
